viewer/packages/tools/src/Measurement/MeasurementTool.ts

```diff
@@ -0,0 +1,215 @@
+```
+ */
+export class MeasurementTool extends Cognite3DViewerToolBase {
+  private _options: Required<MeasurementOptions>;
+
+  private readonly _viewer: Cognite3DViewer;
+  private readonly _geometryGroup = new THREE.Group();
+  private readonly _measurements: Measurement[];
+  private _currentMeasurementIndex: number;
+  private readonly _htmlOverlay: HtmlOverlayTool;
+
+  private readonly _handleLabelClustering = this.createCombineClusterElement.bind(this);
+  private readonly _handlePointerClick = this.onPointerClick.bind(this);
+  private readonly _handlePointerMove = this.onPointerMove.bind(this);
+
+  private readonly _overlayOptions: HtmlOverlayToolOptions = {
+    clusteringOptions: { mode: 'overlapInScreenSpace', createClusterElementCallback: this._handleLabelClustering }
+  };
+
+  private static readonly defaultLineOptions: Required<MeasurementOptions> = {
+    distanceToLabelCallback: d => MeasurementTool.metersLabelCallback(d),
+    lineWidth: 2.0,
+    color: 0x00ffff
+  };
+
+  constructor(viewer: Cognite3DViewer, options?: MeasurementOptions) {
+    super();
+    this._viewer = viewer;
+    this._options = {
+      ...MeasurementTool.defaultLineOptions,
+      ...options
+    };
+    this._measurements = [];
+    this._htmlOverlay = new HtmlOverlayTool(this._viewer, this._overlayOptions);
+    this._currentMeasurementIndex = -1;
+
+    this._geometryGroup.name = MeasurementTool.name;
+    this._viewer.addObject3D(this._geometryGroup);
+  }
+
+  /**
+   * Enter into point to point measurement mode.
+   */
+  enterMeasurementMode(): void {
+    this.setupEventHandling();
+  }
+
+  /**
+   * Exit measurement mode.
+   */
+  exitMeasurementMode(): void {
+    //clear all mesh, geometry & event handling.
+    this._measurements.forEach(measurement => {
+      measurement.dispose();
+    });
+    this.removeEventHandling();
+  }
+
+  /**
+   * Removes a measurement from the Cognite3DViewer.
+   * @param _measurement Measurement mesh to be removed from @Cognite3DViewer.
+   */
+  removeMeasurement(_measurement: THREE.Group): void {
+    // TODO 2022-07-05 larsmoa: Implement
+    throw new Error('Not implemented');
+  }
+
+  /**
+   * Removes all measurements from the Cognite3DViewer.
+   */
+  removeAllMeasurement(): void {
+    this._measurements.forEach(measurement => {
+      measurement.removeMeasurement();
+    });
+    this._measurements.splice(0);
+    this._currentMeasurementIndex = -1;
+  }
+
+  /**
+   * Sets the visiblity of labels in the Measurement.
+   * @param enable
+   */
+  showMeasurementLabels(enable: boolean): void {
+    if (this._htmlOverlay) {
+      this._htmlOverlay.visible(enable);
+    }
+  }
+
+  /**
+   * Sets Measurement line width and color with @options value.
+   * @param options MeasurementOptions to set line width and color.
+   */
+  setLineOptions(options: MeasurementOptions): void {
+    // TODO 2022-07-05 larsmoa: WTF - clean up. Allowing setting options and then ignoring one of the options
```

Comment:
Currently, `lineOptions` are mixed with the callback. Very confusing, can't be only partially used.

Reply:
Partial setting was provided for the user to update only `color` or `lineWidth` or both for the next measuring line.